src/x/sync/pooled_worker_pool.go

```diff
@@ -108,19 +117,29 @@ func (p *pooledWorkerPool) goWithTimeout(work Work, timeout time.Duration) bool
 	}
 
 	if !p.growOnDemand {
-		if timeout <= 0 {
+		if ctx == nil && timeout <= 0 {
 			workCh <- work
 			return true
 		}
 
+		if ctx != nil {
+			// Using context for cancellation not timer.
+			select {
```

Comment:
hmm I wonder if we want to check for cancellation first; seems ineffective to grab a worker if we're already cancelled

Reply:
Sure thing.